### ISSUE TITLE
Cannot union inputs of different types

### ISSUE THREAD

**Issue by RustedBones** [spotify/scio#4494]
In some circumstances, unioning input fails with `Cannot union inputs of different types`.
Types are compared using the coders.

The equality is failing when wrapped and unwrapped materialized coders are compared:
```scala
val coder1: MyCoder
val coder2: MyCoder
println(coder1 == coder2) // true
println(WrappedBCoder(coder1) == coder2) // false
```
This is an issue when union is made on SCollection where one side is wrapped and the other isn't
We should make sure that both sides of SCollection are always wrapped the same way.

**Comment by RustedBones** [comment]
This should have probably been fixed by #4397 since wrapping has been made more systematic at top level

**Comment by RustedBones** [comment]
Happening when we have
```scala
sc: SCollection[MyClass]
sc.coder // LazyCoder(MyClass)
sc.tap(_ => ()).coder // WrappedBCoder(LazyCoder(MyClass))
```
